gotham/src/handler/static_file.rs

```diff
@@ -0,0 +1,892 @@
+        }
+    }
+}
+
+impl DirHandler {
+    /// Create a new `DirHandler` with the given root path.
+    pub fn new<P>(path: P) -> DirHandler
+    where
+        FileOptions: From<P>,
+    {
+        DirHandler {
+            options: FileOptions::from(path),
+        }
+    }
+}
+
+impl NewHandler for FileHandler {
+    type Instance = Self;
+
+    fn new_handler(&self) -> Result<Self::Instance> {
+        Ok(self.clone())
+    }
+}
+
+impl NewHandler for DirHandler {
+    type Instance = Self;
+
+    fn new_handler(&self) -> Result<Self::Instance> {
+        Ok(self.clone())
+    }
+}
+
+impl Handler for DirHandler {
+    fn handle(self, state: State) -> Box<HandlerFuture> {
+        let path = {
+            let mut base_path = PathBuf::from(self.options.path);
+            let file_path = PathBuf::from_iter(&FilePathExtractor::borrow_from(&state).parts);
+            base_path.extend(&normalize_path(&file_path));
+            base_path
+        };
+        create_file_response(
+            FileOptions {
+                path,
+                ..self.options
+            },
+            state,
+        )
+    }
+}
+
+impl Handler for FileHandler {
+    fn handle(self, state: State) -> Box<HandlerFuture> {
+        create_file_response(self.options, state)
+    }
+}
+
+// Creates the `HandlerFuture` response based on the given `FileOptions`.
+fn create_file_response(options: FileOptions, state: State) -> Box<HandlerFuture> {
+    let mime_type = mime_for_path(&options.path);
+    let headers = HeaderMap::borrow_from(&state).clone();
+
+    let (path, encoding) = check_compressed_options(&options, &headers);
+
+    let response_future = File::open(path).and_then(|file| file.metadata()).and_then(
+        move |(file, meta)| {
+            if not_modified(&meta, &headers) {
+                Ok(http::Response::builder()
+                    .status(StatusCode::NOT_MODIFIED)
+                    .body(Body::empty())
+                    .unwrap())
+            } else {
+                let len = meta.len();
+                let buf_size = optimal_buf_size(&meta);
+
+                let stream = file_stream(file, buf_size, len);
+                let body = Body::wrap_stream(stream);
+                let mut response = http::Response::builder();
+                response.status(StatusCode::OK);
+                response.header(CONTENT_LENGTH, len);
+                response.header(CONTENT_TYPE, mime_type.as_ref());
+                response.header(CACHE_CONTROL, options.cache_control);
+
+                if let Some(etag) = entity_tag(&meta) {
+                    response.header(ETAG, etag);
+                }
+                if let Some(content_encoding) = encoding {
+                    response.header(CONTENT_ENCODING, content_encoding);
+                }
+
+                Ok(response.body(body).unwrap())
+            }
+        },
+    );
+    Box::new(response_future.then(|result| match result {
```

Comment:
You're right - but I struggled to make the borrow checker happy about it.
Maybe I'm missing something dumb :)
```
    Box::new(
        response_future
            .map(|response| (state, response))
            .map_err(|e| {
                let status = error_status(&e);
                (state, e.into_handler_error().with_status(status))
            }),
    )
```
```
230 |             .map(|response| (state, response))
    |                  ---------- value moved (into closure) here
...
233 |                 (state, e.into_handler_error().with_status(status))
    |                  ^^^^^ value captured here after move
 = note: move occurs because `state` has type `state::State`, which does not implement the `Copy` trait

```

Reply:
@colinbankier actually, yeah. Makes sense that it's how it is in that case :p